rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,132 @@
+
+### UX
+
+The backend and audit log are configured in the `storage` section of the `teleport.yaml` config file:
+
+```yaml
+teleport:
+  storage:
+    type: postgresql
+    conn_string: host=testdb.postgres.database.azure.com port=5432 sslmode=verify-full user=teleport_auth dbname=teleport_backend
+    auth_mode: azure
+    azure_client_id: ""
+
+    change_feed_poll_interval: "1s"
+    change_feed_batch_size: 10000
+
+    disable_expiry: false
+    expiry_interval: "30s"
+    expiry_batch_size: 1000
+
+    audit_events_uri:
+      - postgresql://teleport_auth@testdb.postgres.database.azure.com:5432/teleport_events?sslmode=verify-full&auth_mode=azure&disable_cleanup=false&cleanup_interval=1h&retention_period=8766h
+```
+
+The `conn_string` used by the backend is a [`libpq`-compatible Postgres connection string](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING), and can be specified in both keyword-value format and as a URI. The audit log URI can only be specified as a URI. In addition to the standard Postgres options, the options supported by [`pgxpool`](https://pkg.go.dev/github.com/jackc/pgx/v5/pgxpool#ParseConfig) are also supported - `pool_max_conns` being the most useful.
+
+The `auth_mode: azure` field for the backend and the `auth_mode=azure` query parameter enable authentication to the database via Azure AD. The optional `azure_client_id` field (for both the backend and the audit log) override the default azure credential selection and the selection in the `AZURE_CLIENT_ID` envvar, so it's possible to select two different managed identities for backend and audit log. A blank (or unset) `auth_mode` will use the default Postgres authentication methods.
+
+The `change_feed_*` options regulate the internal behavior of the change feed polling, the `expiry_*` and `disable_expiry` regulate the behavior of the TTL deletions. We're not expecting users to have to tweak these options (except maybe `disable_expiry` if TTL deletion is going to be run externally with `pg_cron` or similar methods), they're just exposed for performance tuning.
+
+The `disable_cleanup`, `cleanup_interval` and `retention_period` options to the audit log are optional, and can be set and changed to support manual cleanup of old events, or to change the data retention period for the audit log. The default `retention_period` is one year (365.25 days).
+
+### Rejected alternatives and future work
+
+Early tests with an in-database change feed mechanism (using a global, transactional counter to guarantee correct ordering of events in an `events` table) seemed to show that the performance was not up to the task, failing to sustain idle cluster operations for more than 3500 connected agents. It's possible that using some sharded counter would improve the performance of such an approach, but the added complexity in polling for such events makes it a less appealing option than just using logical decoding together with very simple operations on the data.
```

Comment:
Are these 3500 agents connected directly to the DB server? From one of the conversations I had, I took that only the auth service has a direct database access, and all other nodes are talking to the DB through it. Did I get it wrong?

Reply:
That's right, the load test involved connecting a lot of Teleport agents (running the SSH service) to an appropriately-sized Teleport cluster - each agent, at a minimum, does one `KeepAlive` operation per minute to refresh its heartbeat.